### ISSUE TITLE
TLS 1.2 issue on specific devices below API level 21

### ISSUE THREAD

**Issue by bentmar** [Microsoft/AppCenter-SDK-Android#917]
### **Description**

Appcenter fails to send data to server when running kitkat with tls 1.2 not enabled by default

### **Repro Steps**

Use appcenter (analytics + crashes is what I use) on a Xamarin client. 
Run on a KitKat android device that doesnt have tls 1.2 enabled by default
Track some data using appcenter api
See these error logs in debug output:

`
12-18 12:19:05.372 W/AppCenter( 4447): Try #3 failed and will be retried in 1077322 ms
12-18 12:19:05.372 W/AppCenter( 4447): javax.net.ssl.SSLException: SSL handshake aborted: ssl=0x6ec7d2f8: I/O error during system call, Connection reset by peer
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:486)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.Connection.upgradeToTls(Connection.java:146)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.Connection.connect(Connection.java:107)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:294)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.internal.http.HttpEngine.sendSocketRequest(HttpEngine.java:255)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:206)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:345)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:89)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:197)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.android.okhttp.internal.http.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:254)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.microsoft.appcenter.http.DefaultHttpClientCallTask.doHttpCall(DefaultHttpClientCallTask.java:249)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.microsoft.appcenter.http.DefaultHttpClientCallTask.doInBackground(DefaultHttpClientCallTask.java:296)
12-18 12:19:05.372 W/AppCenter( 4447): 	at com.microsoft.appcenter.http.DefaultHttpClientCallTask.doInBackground(DefaultHttpClientCallTask.java:39)
12-18 12:19:05.372 W/AppCenter( 4447): 	at android.os.AsyncTask$2.call(AsyncTask.java:288)
12-18 12:19:05.372 W/AppCenter( 4447): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
12-18 12:19:05.372 W/AppCenter( 4447): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
12-18 12:19:05.372 W/AppCenter( 4447): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
12-18 12:19:05.372 W/AppCenter( 4447): 	at java.lang.Thread.run(Thread.java:841)
`
### **Details**

1. Which SDK version are you using?
     1.12.0 (DotNetClient, Xamarin)
2. Which OS version did you experience the issue on?
    Kitkat 

3. What device version did you see this error on?  Were you using an emulator or a physical device?
    only tested on physical

5. Please enable verbose logging for your app using `AppCenter.setLogLevel(Log.VERBOSE)` before your call to `AppCenter.start(...)` and include the logs here:

`12-18 14:11:09.262 V/AppCenter( 5160): Calling https://in.appcenter.ms/logs?api-version=1.0.0...
12-18 14:11:09.262 V/AppCenter( 5160): Headers: {Install-ID=f5a46158-96bf-4c9f-941e-b178fe8c1b45, Content-Type=application/json, App-Secret=****************************9b232a6e}
12-18 14:11:09.262 V/AppCenter( 5160): {
12-18 14:11:09.262 V/AppCenter( 5160):   "logs": [
12-18 14:11:09.262 V/AppCenter( 5160):     {
12-18 14:11:09.262 V/AppCenter( 5160):       "id": "ff9c63a0-56cc-4e66-8d75-a8b7da5caa17",
12-18 14:11:09.262 V/AppCenter( 5160):       "device": {
12-18 14:11:09.262 V/AppCenter( 5160):         "screenSize": "480x640",
12-18 14:11:09.262 V/AppCenter( 5160):         "model": "MC67NA",
12-18 14:11:09.262 V/AppCenter( 5160):         "osVersion": "4.4.4",
12-18 14:11:09.262 V/AppCenter( 5160):         "timeZoneOffset": 0,
12-18 14:11:09.262 V/AppCenter( 5160):         "appBuild": "132",
12-18 14:11:09.262 V/AppCenter( 5160):         "locale": "en_US",
12-18 14:11:09.262 V/AppCenter( 5160):         "sdkVersion": "1.11.0",
12-18 14:11:09.262 V/AppCenter( 5160):         "oemName": "Zebra Technologies",
12-18 14:11:09.262 V/AppCenter( 5160):         "wrapperSdkVersion": "1.12.0",
12-18 14:11:09.262 V/AppCenter( 5160):         "appVersion": "2.0.0",
12-18 14:11:09.262 V/AppCenter( 5160):         "appNamespace": "com.****.****",
12-18 14:11:09.262 V/AppCenter( 5160):         "wrapperSdkName": "appcenter.xamarin",
12-18 14:11:09.262 V/AppCenter( 5160):         "osApiLevel": 19,
12-18 14:11:09.262 V/AppCenter( 5160):         "osName": "Android",
12-18 14:11:09.262 V/AppCenter( 5160):         "sdkName": "appcenter.android",
12-18 14:11:09.262 V/AppCenter( 5160):         "osBuild": "01-18-01-4AJ22-K-00-M1-082716",
12-18 14:11:09.262 V/AppCenter( 5160):         "wrapperRuntimeVersion": "9.1.0.29"
12-18 14:11:09.262 V/AppCenter( 5160):       },
12-18 14:11:09.262 V/AppCenter( 5160):       "timestamp": "2018-12-18T14:10:58.061Z",
12-18 14:11:09.262 V/AppCenter( 5160):       "sid": "12b8a491-49a9-4256-8a96-e787513556fd",
12-18 14:11:09.262 V/AppCenter( 5160):       "type": "event",
12-18 14:11:09.262 V/AppCenter( 5160):       "name": "***"
12-18 14:11:09.262 V/AppCenter( 5160):     }
12-18 14:11:09.262 V/AppCenter( 5160):   ]
12-18 14:11:09.262 V/AppCenter( 5160): }
12-18 14:11:09.983 W/AppCenter( 5160): Try #2 failed and will be retried in 170993 ms
12-18 14:11:09.983 W/AppCenter( 5160): javax.net.ssl.SSLException: SSL handshake aborted: ssl=0x6cf3d4e0: I/O error during system call, Connection reset by peer
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:486)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.Connection.upgradeToTls(Connection.java:146)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.Connection.connect(Connection.java:107)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:294)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.internal.http.HttpEngine.sendSocketRequest(HttpEngine.java:255)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:206)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:345)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:89)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:197)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.android.okhttp.internal.http.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:254)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.microsoft.appcenter.http.DefaultHttpClientCallTask.doHttpCall(DefaultHttpClientCallTask.java:249)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.microsoft.appcenter.http.DefaultHttpClientCallTask.doInBackground(DefaultHttpClientCallTask.java:296)
12-18 14:11:09.983 W/AppCenter( 5160): 	at com.microsoft.appcenter.http.DefaultHttpClientCallTask.doInBackground(DefaultHttpClientCallTask.java:39)
12-18 14:11:09.983 W/AppCenter( 5160): 	at android.os.AsyncTask$2.call(AsyncTask.java:288)
12-18 14:11:09.983 W/AppCenter( 5160): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
12-18 14:11:09.983 W/AppCenter( 5160): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
12-18 14:11:09.983 W/AppCenter( 5160): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
12-18 14:11:09.983 W/AppCenter( 5160): 	at java.lang.Thread.run(Thread.java:841)`

**Comment by bentmar** [comment]
https://github.com/Microsoft/AppCenter-SDK-Android/blob/8acb92dbe832d1dbd380fd68223bbcfd4adea559/sdk/appcenter/src/main/java/com/microsoft/appcenter/http/DefaultHttpClientCallTask.java#L178

This is what i hoped was missing, but it´s there allright.

**Comment by phdonnelly** [comment]
This also affects the Distribute module; API queries to https://api.appcenter.ms/ fail and update functionality does not work.

**Comment by ElektrojungeMS** [comment]
Thx for reporting this @bentmar @phdonnelly - we're looking at this to make sure it's fixed in the January release.

**Comment by guperrot** [comment]
The bug seems to affect only specific devices like Samsung ones. The current SDK works well on kitkat devices we own (or emulators).

We tried to apply some modifications to the way we configure TLS so that it works for a larger set of older devices.

For Java an internal release can be tested to see if that fixes the issue:

Add this to application build.gradle:

```
repositories {
    maven {
        url  "https://dl.bintray.com/vsappcenter/appcenter-snapshot" 
    }
}
```

And use specifically version `1.11.1-5+f94875e3` not any other version, this one was built from a branch and not develop or master.

For Xamarin I will post link to nugets in https://github.com/Microsoft/AppCenter-SDK-DotNet/issues/795.

**Comment by bentmar** [comment]
The failing device for me is a"Zebra MC67NA". That device doesnt come with GMS (google mobile services). It might be that devices with GMS updates its own tls to default to 1.2 (you can do this update from code) so maybe non GMS devices are a little trickier..

**Comment by guperrot** [comment]
@bentmar even if it's a different brand, have you tested the prerelease SDK with the changes?

**Comment by bentmar** [comment]
@guperrot Its working fine with the new nugets u gave me!

**Comment by guperrot** [comment]
Keeping issue open just temporarily until official jCenter libs have the fix.

**Comment by guperrot** [comment]
This bug has been fixed in version [1.11.2](https://github.com/Microsoft/AppCenter-SDK-Android/releases/tag/1.11.2).

This bug fix will later be included in Xamarin, React Native, Cordova and Unity.
